Sources/NIO/ByteCollectionUtils.swift

```diff
@@ -0,0 +1,48 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2018 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+fileprivate let defaultWhitespaces = [" ", "\t"].map({UInt8($0.utf8CString[0])})
+
+extension Sequence where Self.Element: Equatable {
+    public func dropLast(while: (Self.Element) throws -> (Bool)) rethrows -> Self.SubSequence {
+        let numLastTrimmedElements = try self.reversed().enumerated().first(where: {try !`while`($0.element)})?.offset ?? 0
+        return self.dropLast(numLastTrimmedElements)
+    }
+    
+    public func trim(limitingElements: [Self.Element]) -> Self.SubSequence {
+        return dropLast(while: {limitingElements.contains($0)})
+                  .drop(while: {limitingElements.contains($0)})
+    }
+}
+
+extension Sequence where Self.Element == UInt8 {
+    public var trimSpaces: Self.SubSequence {
+        return trim(limitingElements: defaultWhitespaces)
+    }
+}
```

Comment:
So the three methods here (`dropLast`, `trim`, `trimSpaces`) are all currently written as extensions on `Sequence`, but we only ever use them on `ByteBufferView` objects, which conform to `RandomAccessCollection`. That means we can further reduce the amount of code here by instead implementing `trim` by removing our implementation of `dropLast`, and instead doing this:

```swift
extension ByteBufferView {
    func trim(limitingElements: [Self.Element]) -> Self.Subsequence {
        let lastNonWhitespaceIndex = self.lastIndex { !limitingElements.contains($0) }
        let firstNonWhitespaceIndex = self.firstIndex { !limitingElements.contains($0) }

        return self[firstNonWhitespaceIndex...lastNonWhitespaceIndex]
    }
}
```

This also avoids us doing the dangerous thing of extending a type that we do not control (namely, `Sequence`).